### PR TITLE
Added  interfaces to manage authentication, event and function definition collections

### PR DESCRIPTION
**Pull request by cdavernas**
**Many thanks for submitting your Pull Request :heart:!**

**What this PR does / why we need it**:

Adds  interfaces to manage authentication, event and function definition collections

### CODE REVIEW

File: src/apps/Synapse.Cli/Commands/Workflows/DeployWorkflowCommand.cs [hunk 1]
```diff
@@ -73,7 +73,7 @@ public async Task HandleAsync(string file)
             try
             {
                 workflowDefinition = await this.WorkflowReader.ReadAsync(stream);
-                var workflow = await this.SynapseManagementApi.CreateWorkflowAsync(new() { Definition = workflowDefinition });
+                var workflow = await this.SynapseManagementApi.CreateWorkflowAsync(new() { Collection = workflowDefinition });
```
Comment (JBBianchi) [inline src/apps/Synapse.Cli/Commands/Workflows/DeployWorkflowCommand.cs:76]: `Collection = workflowDefinition` reads a bit weird

File: src/apps/Synapse.Worker/Services/Processors/OpenApiFunctionProcessor.cs [hunk 2]
```diff
@@ -174,6 +174,8 @@ protected override async Task InitializeAsync(CancellationToken cancellationToke
                     .Single(p => p.Value.Operations.Any(o => o.Value.OperationId == operation.Value.OperationId));
                 this.Path = path.Key;
                 await this.BuildParametersAsync(cancellationToken);
+                if (this.Parameters == null)
+                    return;
```
Comment (JBBianchi) [inline src/apps/Synapse.Worker/Services/Processors/OpenApiFunctionProcessor.cs:178]: Ideally, should be on its own PR.

File: src/dashboard/Synapse.Dashboard/Pages/Resources/Collections/Authentications/AuthenticationDefinitionCollectionEditorSelectors.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        public static IObservable<string?> SelectSerializedAuthenticationDefinitionCollection(IStore store)
+        {
+            return store.GetFeature<AuthenticationDefinitionCollectionEditorState>()
+                .Select(featureState => featureState.SerializedCollection)
+                .DistinctUntilChanged();
+        }
+
+    }
+
+}
```
Comment (JBBianchi) [inline src/dashboard/Synapse.Dashboard/Pages/Resources/Collections/Authentications/AuthenticationDefinitionCollectionEditorSelectors.cs:58]: Ideally, add selectors for other state slices like `ExpanderStates`.

File: src/dashboard/Synapse.Dashboard/Pages/Resources/Collections/Authentications/Create.razor [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+                            <tbody>
+                                @if (collection.Authentications != null)
+                                {
+                                    @foreach (var authentication in collection.Authentications)
+                                    {
+                                        <ExpandableDataRow>
+                                            <Header>
+                                            <td>@authentication.Name</td>
+                                            <td><span class="badge bg-secondary">@EnumHelper.Stringify(authentication.Scheme)</span></td>
+                                            <td class="d-flex justify-content-end"><button @onclick="async () => await OnPropertyChanged(nameof(collection.Authentications), d => d.Authentications?.Remove(authentication))" @onclick:stopPropagation="true" class="btn btn-outline-dark"><small class="bi bi-trash"></small></button></td>
```
Comment (JBBianchi) [inline src/dashboard/Synapse.Dashboard/Pages/Resources/Collections/Authentications/Create.razor:96]: Be careful not to [recreate too many delegates](https://learn.microsoft.com/en-us/aspnet/core/blazor/performance?view=aspnetcore-6.0#avoid-recreating-delegates-for-many-repeated-elements-or-components).

File: src/dashboard/Synapse.Dashboard/Pages/Resources/Collections/Authentications/Create.razor [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+    protected virtual void OnToggleExpand(string sectionName, bool isExpanded)
+    {
+        this.Dispatcher.Dispatch(new ToggleExpand(sectionName, isExpanded));
+        this.StateHasChanged();
+    }
+
+    protected virtual async Task OnAutoGenerateIdChanged(bool autoGenerate)
+    {
+        this.autoGenerateId = autoGenerate;
```
Comment (JBBianchi) [inline src/dashboard/Synapse.Dashboard/Pages/Resources/Collections/Authentications/Create.razor:202]: Could be an action.

File: src/dashboard/Synapse.Dashboard/Pages/Resources/Collections/Authentications/Create.razor [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+        if (this.autoGenerateId)
+            await this.OnPropertyChanged(nameof(collection.Name), w => w.Id = w.Name.Slugify("-").ToLowerInvariant());
+    }
```
Comment (JBBianchi) [inline src/dashboard/Synapse.Dashboard/Pages/Resources/Collections/Authentications/Create.razor:205]: Could be an action/effect.

File: src/dashboard/Synapse.Dashboard/Pages/Resources/Collections/Authentications/Create.razor [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+    {
+        if (this.collection == null)
+            return;
+        await this.OnPropertyChanged(nameof(collection.Name), w =>
+        {
+            w.Name = name;
+            if (autoGenerateId && !string.IsNullOrWhiteSpace(name))
+                w.Id = name.Slugify("-").ToLowerInvariant();
+        });
+    }
```
Comment (JBBianchi) [inline src/dashboard/Synapse.Dashboard/Pages/Resources/Collections/Authentications/Create.razor:217]: Could be an action/effect.

File: src/dashboard/Synapse.Dashboard/Pages/Resources/Collections/ResourceListActions.cs [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+        /// <param name="querySetup">An <see cref="Action{T}"/> used to setup the query to perform</param>
+        public QueryV1FunctionDefinitionCollections(string? searchTerm, Action<IODataQueryCollection<V1FunctionDefinitionCollection>> querySetup)
+        {
+            var builder = new ODataQueryBuilder(new Uri("https://test.com"))
+                .For<V1FunctionDefinitionCollection>("V1FunctionDefinitionCollections")
+                .ByList();
+            querySetup(builder);
+            Query = builder.ToUri(UriKind.Absolute).Query;
+            if (!string.IsNullOrWhiteSpace(searchTerm))
+                Query = $"$search={searchTerm}&{Query}";
```
Comment (JBBianchi) [inline src/dashboard/Synapse.Dashboard/Pages/Resources/Collections/ResourceListActions.cs:52]: Missing `this.`

I guess `search` is not supported by the query builder?

File: src/dashboard/Synapse.Dashboard/Pages/Resources/List.razor [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+                if (state.EventDefinitionCollections != null)
+                    this.eventDefinitionCollections = state.EventDefinitionCollections
+                        .GroupBy(c => c.Name)
+                        .ToDictionary(g => g.Key, g => g.OrderByDescending(c => SemVersion.Parse(c.Version, SemVersionStyles.Any)).ToList())!;
+                if (state.AuthenticationDefinitionCollections != null)
+                    this.authenticationDefinitionCollections = state.AuthenticationDefinitionCollections
+                        .GroupBy(c => c.Name)
+                        .ToDictionary(g => g.Key, g => g.OrderByDescending(c => SemVersion.Parse(c.Version, SemVersionStyles.Any)).ToList())!;
+                this.StateHasChanged();
+            });
```
Comment (JBBianchi) [inline src/dashboard/Synapse.Dashboard/Pages/Resources/List.razor:272]: Ideally create selectors.

File: src/dashboard/Synapse.Dashboard/Pages/Workflows/View.razor [hunk 10]
```diff
@@ -156,7 +156,7 @@
             }
             else if(workflowInstance != null)
             {
-                <TabControl ContainerCssClass="mh-100">
+                <TabControl ContainerCssClass="mh-100 d-flex flex-column">
```
Comment (JBBianchi) [inline src/dashboard/Synapse.Dashboard/Pages/Workflows/View.razor:159]: Missing class `container` ?

**Review by JBBianchi — APPROVED** [review]
Some comments but nothing blocking.

File: src/apps/Synapse.Cli/Commands/Workflows/DeployWorkflowCommand.cs (hunk 1 above)
Comment (cdavernas) [reply src/apps/Synapse.Cli/Commands/Workflows/DeployWorkflowCommand.cs:76]: It's a bug. Fixing

File: src/apps/Synapse.Worker/Services/Processors/OpenApiFunctionProcessor.cs (hunk 2 above)
Comment (cdavernas) [reply src/apps/Synapse.Worker/Services/Processors/OpenApiFunctionProcessor.cs]: I created, pushed and merged https://github.com/serverlessworkflow/synapse/pull/278 for that specific fix. Thanks!

File: src/apps/Synapse.Cli/Commands/Workflows/DeployWorkflowCommand.cs (hunk 1 above)
Comment (cdavernas) [reply src/apps/Synapse.Cli/Commands/Workflows/DeployWorkflowCommand.cs]: Fixed by https://github.com/serverlessworkflow/synapse/pull/277/commits/0f46de36fe78826642ae487eb7ca8cf6f1cc068f

File: src/dashboard/Synapse.Dashboard/Pages/Workflows/View.razor (hunk 10 above)
Comment (cdavernas) [reply src/dashboard/Synapse.Dashboard/Pages/Workflows/View.razor]: Fixed by https://github.com/serverlessworkflow/synapse/pull/277/commits/ef586f5fcc591112b223e41d64a77288e377794c
